drivers/mtd/mtd.c

```diff
@@ -56,12 +102,57 @@ int mtd_write(mtd_dev_t *mtd, const void *src, uint32_t addr, uint32_t count)
         return -ENODEV;
     }
 
-    if (mtd->driver->write) {
-        return mtd->driver->write(mtd, src, addr, count);
+    /* page size is always a power of two */
+    const uint32_t page_shift = 32 - __builtin_clz(mtd->page_size) - 1;
+    const uint32_t page_mask = mtd->page_size - 1;
+
+    return mtd_write_page(mtd, src, addr >> page_shift, addr & page_mask, count);
+}
+
+int mtd_write_page(mtd_dev_t *mtd, const void *src, uint32_t page, uint32_t offset,
+                   uint32_t count)
+{
+    if (!mtd || !mtd->driver) {
+        return -ENODEV;
     }
-    else {
-        return -ENOTSUP;
+
+    if (mtd->driver->write_page == NULL) {
+        /* TODO: remove when all backends implement write_page */
+        if (mtd->driver->write) {
+            return mtd->driver->write(mtd, src, mtd->page_size * page + offset, count);
+        } else {
+            return -ENOTSUP;
+        }
+    }
+
+    const char *_src = src;
+
+    /* page size is always a power of two */
+    const uint32_t page_shift = 32 - __builtin_clz(mtd->page_size) - 1;
+    const uint32_t page_mask = mtd->page_size - 1;
+
+    page  += offset >> page_shift;
+    offset = offset & page_mask;
+
+    while (count) {
+        int written = mtd->driver->write_page(mtd, _src, page, offset, count);
```

Comment:
Same, `mtd_desc::write_page` should return 0 on success, not the number of bytes written.